include/networkit/io/DotGraphWriter.hpp

```diff
@@ -18,16 +23,16 @@ namespace NetworKit {
  * If a more thorough support is desired, please contact the developers over networkit@ira.uni-karlsruhe.de.
  *
  * [1] http://www.graphviz.org/Documentation/dotguide.pdf
- */ 
-class DotGraphWriter {
+ */
+class DotGraphWriter final : public GraphWriter {
 public:
 	/**
 	 * Write a graph as a GraphViz/file.
-	 * 
+	 *
 	 * @param[in]	graph	The graph object
 	 * @param[in]	path	The file path to be written to
 	 */
-	virtual void write(Graph& graph, std::string path) const;
```

Comment:
Good point, as you wrote those classes could easily be adapted to implement a const `write` method.

Reply:
I disagree here. Why `write()` is essentially the `GraphWriter`s `run` method. If you want to write in parallel, construct two writers. Making the virtual method `const` heavily constraints what is efficiently possible in a `GraphWriter`.

Reply:
To elaborate: The base class for `GraphWriter`s should absolutely remain non-`const`. I don't see what advantage `const` offers for writers.

More broadly, I do not think that we should support parallel usage of *any* algorithm (or I/O interface) in NetworKit *unless* there is a good reason to do so. `const` is an unnecessary pessimization here: it forces implementers to maintain local copies of algorithm state instead of storing the state in the algorithm's object. Often, this is much less efficient, especially if an algorithm's object is reused. Reuse of algorithm objects is arguably much more common than parallel use of the same algorithm's object.

Also note that `const` does affect the API: turning a `const` method to non-`const` is a breaking change. We should only apply `const` to functions that will not write to the algorithm's state, forever, regardless of implementation strategy. I am not sure if *any* of the `GraphWriter`s should have `const write()` methods.

Reply:
While I see your point and are kind of convinced that the base class should remain non-const, I still have two points:
- Currently there ARE writers that are const and the breaking of API you mentioned would occur in this PR
- AFAICS we currently have no examples where the state of the write should change (I do not think that `nodes` should be a member of `NetworKitBinaryWriter` but rather a local variable; it's only used in `write`. In `GraphToolBinaryWriter` no state is changed)

Reply:
I agree that those `count` variables should be local. However, there might be writers that have a legitimate reason to construct large data structures (e.g., `std::vector`s or `std::unordered_map`s to remap node IDs?) in memory. Those structures should probably be stored in the object to avoid reallocating them whenever `write()` is called.

Reply:
Yes, I agree that for future writers it might be an issue, and -- given that -- I agree that the base class should remain non-const. I'm just saying that if we do this, we should make a conscious decision to break the interface (also as you already pointed out with probably no serious implications).

Reply:
Now the base class `write` is non-const. Are there any other issues or can we merge this?